beets/ui/__init__.py

```diff
@@ -1057,43 +1048,149 @@ def _open_library(config):
     return lib
 
 
-def _raw_main(args, lib=None):
+class BeetsCLI(click.MultiCommand):
+
+    def _commands(self, ctx):
+        from beets.ui.commands import default_commands
+
+        config = _configure(ctx.params)
+        _load_plugins(config)
+
+        subcommands = list(default_commands)
+        subcommands.extend(plugins.commands())
+
+        return {
+            cmd.name: cmd
+            for cmd in subcommands
+            if isinstance(cmd, click.Command)
+        }
+
+    def list_commands(self, ctx):
+        return self._commands(ctx).keys()
+
+    def get_command(self, ctx, name):
+        # Special case for the `config --edit` command: bypass loading config
+        # so that an invalid configuration does not prevent the editor from
+        # starting.
+        if ctx.args and ctx.args[0] == 'config' and \
+           ('-e' in ctx.args or '--edit' in ctx.args):
+            from beets.ui.commands import config_cmd
+            return config_cmd
+
+        return self._commands(ctx).get(name)
```

Comment:
You'd still need to load the config before, to get the active plugins and their subcommands.